test/src/test/java/hudson/model/QueueTest.java

```diff
@@ -780,4 +785,85 @@ public void run() {
                 "B finished at " + buildBEndTime + ", C started at " + buildC.getStartTimeInMillis(), 
                 buildC.getStartTimeInMillis() >= buildBEndTime);
     }
+
+    @Issue("JENKINS-30084")
+    @Test
+    /*
+     * When a flyweight task is restricted to run on a specific node, the node will be provisioned
+     * and the flyweight task will be executed.
+     */
+    public void shouldRunFlyweightTaskOnProvisionedNodeWhenNodeRestricted() throws Exception {
+        MatrixProject matrixProject = r.createMatrixProject();
+        matrixProject.setAxes(new AxisList(
+                new Axis("axis", "a", "b")
+        ));
+        Label label = LabelExpression.get("aws-linux-dummy");
+        DummyCloudImpl dummyCloud = new DummyCloudImpl(r, 0);
+        dummyCloud.label = label;
+        r.jenkins.clouds.add(dummyCloud);
+        matrixProject.setAssignedLabel(label);
+        r.assertBuildStatusSuccess(matrixProject.scheduleBuild2(0));
+        assertEquals("aws-linux-dummy", matrixProject.getBuilds().getLastBuild().getBuiltOn().getLabelString());
+    }
+
+    @Test
+    public void shouldBeAbleToBlockFlyWeightTaskOnLastMinute() throws Exception {
+        MatrixProject matrixProject = r.createMatrixProject("downstream");
+        matrixProject.setAxes(new AxisList(
+                new Axis("axis", "a", "b")
+        ));
+        Label label = LabelExpression.get("aws-linux-dummy");
+        DummyCloudImpl dummyCloud = new DummyCloudImpl(r, 0);
+        dummyCloud.label = label;
+        PropertyImpl property = new PropertyImpl();
+        dummyCloud.getNodeProperties().add(property);
+        r.jenkins.clouds.add(dummyCloud);
+        matrixProject.setAssignedLabel(label);
+        FreeStyleProject upstreamProject = r.createFreeStyleProject("upstream");
+        //let's assume the flyweighttask has an upstream project and that must be blocked
+        // when the upstream project is running
+        matrixProject.addTrigger(new ReverseBuildTrigger("upstream", Result.SUCCESS));
+        matrixProject.setBlockBuildWhenUpstreamBuilding(true);
+        //we schedule the project but we pretend no executors are available thus
+        //the flyweight task is in the buildable queue without being executed
+        matrixProject.scheduleBuild2(0);
+        //in this state the build is not blocked, it's just waiting for an available executor
+        assertFalse(matrixProject.isBuildBlocked());
+        //we start the upstream project that should block the downstream one
+        upstreamProject.getBuildersList().add(new SleepBuilder(10000));
+        QueueTaskFuture upstream = upstreamProject.scheduleBuild2(0);
+        upstream.waitForStart();
+        //let's wait for the Queue to be updated
+        Thread.sleep(5000);
```

Comment:
Presence of a long `Thread.sleep` (or `SleepBuilder`) in a test generally means the test is not reliable. There are various ways to write it better, such as using `Semaphore`s, or waiting _briefly_ in a loop such as `JenkinsRule.waitUntilNoActivity` does.